### PR TITLE
Fixes simplify/doit methods for Summations and Products

### LINKED ISSUE #22260: Simplify should simplify argument of Sum and Product
```
In [20]: from sympy import *

In [21]: n = Symbol('n', integer=True)

In [22]: Sum(sin(n)**2 + cos(n)**2 - 1, (n, 1, oo)).simplify()
Out[22]: 
    ∞         ∞                      
   ___       ___                     
   ╲         ╲                       
    ╲         ╲   ⎛   2         2   ⎞
-   ╱   1 +   ╱   ⎝sin (n) + cos (n)⎠
   ╱         ╱                       
   ‾‾‾       ‾‾‾                     
  n = 1     n = 1                    
```

Hence, it would make sense to apply `simplify` on the argument first like `Integral` seems to do:
```
In [26]: Integral(sin(x)**2 + cos(x)**2 - 1, (x, 1, oo)).simplify()
Out[26]: 0
```

### PR DESCRIPTION
**Pull request by anutosh491**
#### References to other Issues or PRs

Fixes #22260 
In relation to #22200
#### Brief description of what is fixed or changed
As shown in the referenced issue simplify work for integrals and not for Sum/Product and obviously this is because some expressions need evaluation/manipulations or in better terms simplifications to return the expected result .
`_eval_simplify` method of `integrals.py` contains something like `expr.simplify(**kwargs)` which does the needful for integrals , hence I have added the simplify method similarly at appropriate positions .
In the process I also realized that the `doit` method in product has to be fixed for `simplify` to work properly and `doit `method calls `eval_product ` where I've made the relevant changes . 
Changes from current master
Master:
```
>>> Product(sin(n)**2 + cos(n)**2 - 1, (n, 1, 6)).simplify()
(-1 + sin(3)**2 + cos(3)**2)*(-1 + sin(6)**2 + cos(6)**2)*(-1 + cos(1)**2 + sin(1)**2)*(-1 + cos(2)**2 + sin(2)**2)*(-1 + cos(4)**2 + sin(4)**2)*(-1 + cos(5)**2 + sin(5)**2)

>>> Product(sin(n)**2 + cos(n)**2 , (n, 1, 6)).simplify()
(sin(3)**2 + cos(3)**2)*(sin(6)**2 + cos(6)**2)*(cos(1)**2 + sin(1)**2)*(cos(2)**2 + sin(2)**2)*(cos(4)**2 + sin(4)**2)*(cos(5)**2 + sin(5)**2)

>>> Product(sin(n)**2 + cos(n)**2 - 1, (n, 1, oo)).simplify()
Product(sin(n)**2 + cos(n)**2 - 1, (n, 1, oo))
>>>Product(sin(n)**2 + cos(n)**2 + 1, (n, 1, oo)).simplify()
Product(sin(n)**2 + cos(n)**2 + 1, (n, 1, oo))

>>> Sum(sin(n)**2 + cos(n)**2 - 1, (n, 1, oo)).simplify() # This currently takes over 8-10 seconds to calculate
-Sum(1, (n, 1, oo)) + Sum(sin(n)**2 + cos(n)**2, (n, 1, oo))
```
Committed branch:
```
>>> Sum(sin(n)**2 + cos(n)**2 - 1, (n, 1, oo)).simplify() # works much much faster !
0
>>> Sum(sin(n)**2 + cos(n)**2 - 1, (n, 1, 5)).simplify()
0
>>> Product(sin(n)**2 + cos(n)**2 - 1, (n, 1, oo)).simplify()
0
>>> Product(sin(n)**2 + cos(n)**2 - 1, (n, 1, 5)).simplify()
0
>>> Product(sin(n)**2 + cos(n)**2, (n, 1, 10)).simplify()
1
>>> Product(sin(n)**2 + cos(n)**2, (n, 1, oo)).simplify()
1
>>> Product(sin(n)**2 + cos(n)**2 + 1, (n, 1, oo)).simplify()
oo
```

#### Other comments
I have few things in mind where we could improve sympy , will try to implement that in further commits and would also add relevant test cases .Thank You !

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by anutosh491** [comment]
> I'd expect that just simplifying the arguments should take it quite far.


> I'd expect that just simplifying the arguments should take it quite far.

yeah yeah , I haven't been trying to do anything fancy , trying to keep it as simple as possible . Like in `summations.py` I've added a single line which I felt should have been enough . All but 1 test case is not running fine where this line converts a `factorial` into a `gamma ` . Not sure how should I go about addressing this . 
@@ -351,7 +351,7 @@ def _eval_simplify(self, **kwargs):
```

         # next try to combine any interior sums for further simplification
         result = Add(sum_combine(s_t), *o_t)
-
+         result = result.simplify(**kwargs)
         return factor_sum(result, limits=self.limits)

     def is_convergent(self):
```

**Comment by anutosh491** [comment]
> Well, gamma is a generalization of factorial

Well I feel the only test which is not passing in the stats module should be changed. This is because :
```
# Current case 
    X = Poisson('X', 1)
    Y = Poisson('Y', X)
    Z = Poisson('Z', Y)
    exprd = exp(-1)*Sum(exp(-Y)*Y**x*Sum(exp(-X)*X**Y/(factorial(X)*factorial(Y)), (X, 0, oo)), (Y, 0, oo))/factorial(x)
    assert density(Z)(x).simplify() == exprd

# After applying the above simplification method you suggested or even through what I did 
>>> density(Z)(x)
Sum(exp(-Y)*Y**x*Sum(exp(-1)*exp(-X)*X**Y/(factorial(X)*factorial(Y)), (X, 0, oo))/factorial(x), (Y, 0, oo))

>>> density(Z)(x).simplify()
exp(-1)*Sum(exp(-X)*X**Y/gamma(X + 1), (X, 0, oo))*Sum(exp(-Y)*Y**x/gamma(Y + 1), (Y, 0, oo))/gamma(x + 1)

>>> exprd.simplify()
exp(-1)*Sum(exp(-X)*X**Y/gamma(X + 1), (X, 0, oo))*Sum(exp(-Y)*Y**x/gamma(Y + 1), (Y, 0, oo))/gamma(x + 1)
```
Now that we simplify `self.function` inside the `_eval_simplify` we should be equating `density(Z)(x).simplify()` with `exprd.simplify()` , and not `density(Z)(x).simplify()` with `exprd `

**Comment by anutosh491** [comment]
> See #22278 which is an option for the `simplify`-bit (I had to dig into it and noted some things that needed to be changed in general). However, your PR applies `simplify` as part of `doit`, so it can still be useful.
Yeah I went through it , just wanted to ask whether I should be touching the `_eval_simplify()` in `summations.py `block while making my next commit cause we've almost made the same changes as you suggested above.
which are these :
```
        if kwargs.get('deep', True):
            function = self.function.simplify()
        else:
            function = self.function

        # split the function into adds
        terms = Add.make_args(expand(function))
```
I was about to commit this but then I guess this would leave to merge conflicts and I haven't yet been in that situation yet hence thought of asking you !

**Comment by anutosh491** [comment]
Ping @oscargus for review!

**Comment by anutosh491** [comment]
ping @oscargus @smichr for review !

File: sympy/concrete/products.py [hunk 1]
```diff
@@ -332,6 +333,14 @@ def _eval_product(self, term, limits):
 
         elif term.is_Add:
             factored = factor_terms(term, fraction=True)
+            if isinstance(factored.simplify(), Number):
+                factored_simplified = factored.simplify()
```
[the remark this answers is not in the document]
Comment (oscargus) [reply sympy/concrete/products.py:341]: The point here was that you still do simplify twice.

**Comment by oscargus** [comment]
In general, I am a bit doubtful if simplify should be part of doit (it is the other way around, which is maybe also doubtful) and especially interleaving these are not obviously a good idea from a performance perspective.

On the other hand, it is maybe a bit non-obvious that one may have to simplify before doit. But maybe still makes more sense as not all results are simplified etc, so one may sometimes have to apply simplify before one can continue.

In general, one should try to avoid using simplify in library code (unless, of course, in the `_eval_simplify` method).

**Comment by oscargus** [comment]
I realize that this is a bit contradictory to an earlier comment. But it has become clear to me recently, that we rather should try to remove simplify-calls from non-simplify-code to improve the general performance of SymPy, rather than the opposite.

**Comment by anutosh491** [comment]
> I realize that this is a bit contradictory to an earlier comment. But it has become clear to me recently, that we rather should try to remove simplify-calls from non-simplify-code to improve the general performance of SymPy, rather than the opposite.

True , I understand that simplify-calls do hinder the performance . It makes me doubtful of what should we be expecting from this pr then ?? Though I think some cases should pretty much be solved /simplified before returning like the following 

```
>>> Product(sin(n)**2 + cos(n)**2 + 1, (n, 1, 10)).doit()
(sin(3)**2 + cos(3)**2 + 1)*(sin(6)**2 + cos(6)**2 + 1)*(sin(7)**2 + cos(7)**2 + 1)*(sin(9)**2 + cos(9)**2 + 1)*(sin(10)**2 + cos(10)**2 + 1)*(cos(1)**2 + sin(1)**2 + 1)*(cos(2)**2 + sin(2)**2 + 1)*(cos(4)**2 + sin(4)**2 + 1)*(cos(5)**2 + sin(5)**2 + 1)*(cos(8)**2 + sin(8)**2 + 1)
```
Simply giving out 1024 makes much more sense to a user . Though I'll get more clarity on what `doit` is expected to do and maybe not mix the two and expect `doit` to do the simplifications too !

**Comment by anutosh491** [comment]
> In general, I am a bit doubtful if simplify should be part of doit (it is the other way around, which is maybe also doubtful) and especially interleaving these are not obviously a good idea from a performance perspective.
> 
> On the other hand, it is maybe a bit non-obvious that one may have to simplify before doit. But maybe still makes more sense as not all results are simplified etc, so one may sometimes have to apply simplify before one can continue.
> 
> In general, one should try to avoid using simplify in library code (unless, of course, in the `_eval_simplify` method). Edit: the reason is that simplify in itself is quite time consuming, so one do not want to run it arbitrarily.

Maybe this is not the correct way to approach the issue then , i'll try to get some clarity on a possible fix and maybe close this Pr if I can't get anything substantial . Thanks for the reviews !

**Comment by anutosh491** [comment]
Closing because this might not be the correct way to fix mentioned issue . Thanks for the reviews !

**Pull request closed without merging**